### ISSUE TITLE
Add a way for `callbacks` to get `trainer` handler

### ISSUE THREAD

**Issue by MinzhiYoyo** [huggingface/transformers#38181]
When I want to implement differential privacy for the model, I customize the gradient clipping before `optimizer.step()`. The add custom noise to the model after `optimizer.step()`. I cannot get `Trainer.optimizer` in the `callback` function, it shows as `None`. Is it possible to get the reference of `Trainer` directly in `callback`?
